src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -69,31 +97,33 @@ internal sealed class SourceGeneratedDocumentState : DocumentState
         ConstantTextAndVersionSource textSource,
         SourceText text,
         LoadTextOptions loadTextOptions,
-        AsyncLazy<TreeAndVersion> treeSource)
+        AsyncLazy<TreeAndVersion> treeSource,
+        Lazy<Checksum> lazyTextChecksum)
         : base(languageServices, documentServiceProvider, attributes, options, textSource, loadTextOptions, treeSource)
     {
         Identity = documentIdentity;
 
         SourceText = text;
-
-        _lazyTextChecksum = new Lazy<Checksum>(() => Checksum.From(SourceText.GetChecksum()));
+        _lazyTextChecksum = lazyTextChecksum;
     }
 
+    private static Checksum ComputeChecksum(SourceText text)
+        => Checksum.From(text.GetChecksum());
+
     // The base allows for parse options to be null for non-C#/VB languages, but we'll always have parse options
     public new ParseOptions ParseOptions => base.ParseOptions!;
 
-    public Checksum GetTextChecksum()
-        => _lazyTextChecksum.Value;
-
     public SourceGeneratedDocumentContentIdentity GetContentIdentity()
-        => new(this.GetTextChecksum(), this.SourceText.Encoding?.WebName, this.SourceText.ChecksumAlgorithm);
+        => new(this.GetOriginalSourceTextChecksum(), this.SourceText.Encoding?.WebName, this.SourceText.ChecksumAlgorithm);
 
     protected override TextDocumentState UpdateText(ITextAndVersionSource newTextSource, PreservationMode mode, bool incremental)
         => throw new NotSupportedException(WorkspacesResources.The_contents_of_a_SourceGeneratedDocument_may_not_be_changed);
 
-    public SourceGeneratedDocumentState WithUpdatedGeneratedContent(SourceText sourceText, ParseOptions parseOptions)
+    public SourceGeneratedDocumentState WithUpdatedGeneratedContent(
+        SourceText sourceText, ParseOptions parseOptions)
```

Comment:
>SourceText sourceText, ParseOptions parseOptions) [](http://example.com/codeflow?start=8&length=49)

Should this give the option of passing in a Checksum also?

Reply:
Not needed. This is only called from paths on the same side.

Reply:
I believe Cyrus is correct. This is called on the side actually running generators (OOP in this case), so recomputing the checksum would be appropriate in that case when a new text is presented.